### PR TITLE
Expanding the Credits

### PR DESCRIPTION
**Pull request by Zitchas**
Given that this is a community project that is being run, built, and created by the community; I think it is time that we expanded the credits to demonstrate how many more people are involved in this project. Likewise, I think it might be  encouraging to the community at large for people who contribute to actually get their names in here as well.

So, this is just a draft right now. Some of the sections are woefully short on names, but I think this is a good spot to start.

Adding:
- Renaming MZ's section from "programming" to "Creator", because he did so much more than "just" programming.
- Developer section (this is to highlight the fact that we have three "new" developers taking the lead on this project)
- Programming section (Intended to highlight people who have made significant contributions to the engine itself and supporting code.)
- Special Thanks to MCO for the nightlies (fairly self-explanatory)
- Proof Readers & Editors section (There are a number of people who provide huge amounts of feedback making sure we keep our stories making sense, legible, and catching mistakes. I think they deserve to be highlighted)
- Mission & Story Creators (Just like the graphics section highlights people who have made significant contributions to the story, lore, and writing.)
- Thank you to Github Contributers (this is basically a list of everyone who has contributed directly to the game. The only people who have made direct contributions that are not on the list are our four developers. I have sorted the list alphabetically because numbers change. Someone who has contributed a lot up to now, might never contribute again. So I don't see any fair (or reasonable) way to sort based on contributions that wouldn't turn this list into some kind of scoreboard, which isn't the point. And it'd have to be resorted on a regular basis. The point is to recognize everyone who has contributed - whether it is one spelling mistake or entire campaigns.)

Request:
 - Please let me know (preferably by DM on discord to make them easier to keep track of), but here works too) people that you think should be on here. I'm looking especially at the "Proof Readers & Editors" section and the "Mission Writers" section. And I apologize in advance for leaving people out. My mind is fuzzy right now, and world-events are causing a bit of stress, so I know I'm forgetting a lot of people who have provided a lot of help getting our content into a game-worthy state.

### CODE REVIEW

File: credits.txt [hunk 1]
```diff
@@ -61,3 +66,117 @@ Beta Testers
...
+  TheMarksman
+
+Mission & Story Creators
+  Amazinite
+  Zitchas
+
+Thank you to all the following
+people on GitHub who have
+contributed to improving Endless
+Sky:
```
Comment (Amazinite) [inline credits.txt]: The home page of the repo says 151 contributors, but the contributors page only shows 100 (and here we have listed 96, the four developers mentioned as being excluded). Anyone know how to get the remaining 51 (possibly without needing to go through every PR)?

File: credits.txt [hunk 2]
```diff
@@ -11,6 +11,11 @@ Discord (invite: ZeuASSx)
 
 Programming
```
Comment (Amazinite) [inline credits.txt]: Might be worth renaming this to "Creator".

File: credits.txt [hunk 3]
```diff
@@ -11,6 +11,11 @@ Discord (invite: ZeuASSx)
...
   Michael Zahniser
+  
+Developers
+  Amazinite
+  PointedStick
```
Comment (Amazinite) [inline credits.txt]:
```suggestion
  Pointedstick
```

File: credits.txt [hunk 4]
```diff
@@ -11,6 +11,11 @@ Discord (invite: ZeuASSx)
...
+  Tehhowch
```
Comment (Amazinite) [inline credits.txt]:
```suggestion
  tehhowch
```

File: credits.txt [hunk 5]
```diff
@@ -61,3 +66,117 @@ Beta Testers
...
+
+Special thanks to MCOfficer for
+creating, managing, and hosting
+the nightly builds that makes
+so much of this work possible.
+
+Proof Readers & Editors
+  TheMarksman
+
+Mission & Story Creators
```
Comment (Amazinite) [inline credits.txt]: Becca for the Ka'het and a boarding mission. (#4716, #4735)
Arachi and Jana for Nanachi. (#4685)
Pointedstick for various jobs and smaller missions. (#1777, #3235, #3347, #3359)
Unfettered One for various spaceport mission. (#3151, had a hand in #3298, #3781)
MessyMix for an additional mission during the Wanderer campaign. (#2427)
Petervdmeer for the Hauler IV missions. (#4446)
Fzzr for a mission where you capture a smuggler. (#3475)
Anarchist2 for a Wanderer mission. (#4762)
Czartrak for some pirate jobs. (#3984)

File: credits.txt (hunk 2 above)
Comment (Zitchas) [reply credits.txt]: Changed. Thanks!

File: credits.txt (hunk 1 above)
Comment (Zitchas) [reply credits.txt]: Thanks! I've updated the list.

File: credits.txt (hunk 5 above)
Comment (Zitchas) [reply credits.txt]: Thanks, I've added all those in.

**Comment by Amazinite** [comment]
In changing the current "Programming" header to "Creator" since it only had MZ under it, adding a "Programming" header elsewhere with those who have helped make source code changes makes sense to me. MZ, tehhowch, myself, Peter, bojidar, Pointed, Marceline-VQ, ChamEV, Hadron1776, flaviojs, comnom, Fzzr are names that come to mind, although I'm sure there's more.

**Comment by Zitchas** [comment]
It is my hope that this is a living thing that will continue to see people added to it over time, such that it continues to represent the diverse community that has helped shaped the story and world we all enjoy spending time in. So don't consider this PR the end of it. As more stuff is added down the road, this file should be added to as well.

All lists of names that I have added are sorted alphabetically.

**Comment by Zitchas** [comment]
OK, I got the programming credits in there. 

How's it look now?

File: credits.txt [hunk 6]
```diff
@@ -61,3 +79,188 @@ Beta Testers
   Sean Fahey
   Mauricio Gomes
   Iaz Poolar
+
+Special thanks to MCOfficer for
+creating, managing, and hosting
+the nightly builds that makes
+so much of this work possible.
```
Comment (MCOfficer) [inline credits.txt]: not sure if i deserve this special spot, so i'll put it up for discussion. be honest :)

Comment (Zitchas) [reply credits.txt]: Well, obviously, I do. Some of us can do our own builds, but for everyone else, you are solely responsible for allowing everyone to test changes to the engine. There's a fair number of us who *can* do our own builds that simply prefer the convenience of downloading the latest nightly anyway.

Not to mention, didn't you do a whole pile of stuff setting up testing stuff on Github? And then there is James, who is incredibly useful in facilitating content creation discussion on the discord.

In my books, the nightly alone is worth a special thanks. The rest is just gravy on top of that. 

And as an aside, looking through the game credits for things like Mass Effect, people are putting their favourite catering companies in there and thanks to their pets and other stuff. That might be a tad overboard, but compared to that, your contributions are right up there as one of the Devs. Just for the record, I *don't* want us to start including catering companies and pets and whatnot in the credits, but if someone is a really valuable member of the community making a crucial tool to help us build the game... I think they should be in here somehow.

File: credits.txt [hunk 7]
```diff
@@ -61,3 +79,188 @@ Beta Testers
...
+  MasterOfGrey
+  MessyMix
+  Petervdmeer
+  Pointedstick
+  UnfetteredOne
+  Zitchas
+
+Thank you to all the following
+people on GitHub who have
+contributed to Endless-Sky:
```
Comment (petervdmeer) [inline credits.txt:119]: Quite some people in this section are already mentioned earlier under Programming, Proof Readers or Story Creators. I would suggest not repeating them here, since they are already mentioned in more specific sections.

Comment (Zitchas) [reply credits.txt:119]: I left them in this section for clarity, inclusiveness, and simplicity: 
- Some people on the specific lists may want to use an alternative name.
- The list of GitHub contributors is long, and likely to get longer.
- The longer the list is, the more likely we are to have similarly named contributors.
- If the GitHub Contributors list is a straight forward list of everyone who has contributed, then we never have to worry about trying to curate it or filter out people. (whose names may or may not actually be matches for the names in the specific lists).

I initially left the Devs out of it because they are easy to identify, but I am fond of the idea of having one big non-curated list that includes all GitHub contributors specifically so we never have to risk excluding someone by accidentally omitting them because they have a name similar to that of one of the people called out in the curated section.

File: credits.txt (hunk 6 above)
Comment (MCOfficer) [reply credits.txt:86]: i bow to the majority. thank you.

File: credits.txt (hunk 7 above)
Comment (tehhowch) [reply credits.txt:119]: In that case this should be programmatically generated from the repo's commit history or contributor graph. I don't recall if the github display correctly handles now-deactivated accounts (i.e. @ "ghost"), so if it does not then we must parse the commit graph.

Comment (Zitchas) [reply credits.txt:119]: I made the "GitHub Contributors" list the hard way, with Amazinite's help pulling to get the contributers beyond the 100th spot. If there's an automatic way to pull the list and/or update it in the future, sounds great! I don't really want to distract you and our other coder herders from doing game-specific stuff to work on something to parse the script, though. We've got a list now; it shouldn't be too hard to maintain it.

Or were you referring to catching all the people that have made commits/suggestions directly to PRs and thus aren't registered as having committed to the game itself?

Comment (tehhowch) [reply credits.txt:119]: My suggestion is to use a programmatic solution such as the GitHub API (which probably won't handle "ghost") or git commands with some other language (bash, JavaScript, Python) to handle author aggregation and formatting.

Manual parsing of this information is an untenable chore.

Comment (MCOfficer) [reply credits.txt:119]: i tried https://github.com/pydanny/contributors and immediately ran into the ratelimit, no surprise. i could make it work, though, i'm pretty sure.

once we're at it, we can easily set up an action that creates PRs with acknowledgments for new contributors. which reminds me, that's already a thing: https://allcontributors.org/

Comment (Zitchas) [reply credits.txt:119]: That sounds pretty good.

As far as that goes, how about getting the list in its current form merged for the 0.9.12 so that people can see that it has truly become a community project; and have an automatic system for adding people to it in the future be something that is developed for 0.9.13?

Comment (MCOfficer) [reply credits.txt:119]: agreed, moved to #4970

File: credits.txt [hunk 8]
```diff
@@ -61,3 +79,188 @@ Beta Testers
   Sean Fahey
   Mauricio Gomes
   Iaz Poolar
+
+Special thanks to MCOfficer for
+creating, managing, and hosting
+the nightly builds that makes
```
Comment (MCOfficer) [inline credits.txt:85]:
```suggestion
the nightly builds that make
```

File: credits.txt (hunk 7 above)
Comment (MCOfficer) [reply credits.txt:119]: welp, just as i say it, it works. oneliner to get all contributors:
```
$ hub api repos/endless-sky/endless-sky/contributors --paginate | jq '.[].login' -r 
```

here's the output for your convenience: https://gist.github.com/MCOfficer/9b4710251d8ff02a41c5dc52401ed422

File: credits.txt [hunk 9]
```diff
@@ -9,8 +9,26 @@ https://endless-sky.github.io
...
-Programming
+Creator
   Michael Zahniser
+  
+Developers
+  Amazinite
+  Pointedstick
+  tehhowch
+
+Programming
```
Comment (Amazinite) [inline credits.txt:20]: <details><summary>Here's everyone who have made any changes to the /source/ files. I think we should decide what we want to consider "Programming" though, as I wouldn't necessarily call all changes to the source files programming contributions (e.g. typo fixes in comments). There are also changes not made to the source files that should probably be considered programming (e.g. updating the xCode files).</summary>

abenkovskii
AdamKauffman
akien-mga
Amazinite
AMDmi3
bojidar-bg
ChamEV
colbygallup
comnom
CyberShadow
db47h
Disiuze
dplepage
DrBlight
dreness
dzhu
Elyssaen
EndrosG
fcfort
flaviojs
Fzzr
guillochon
Hacklin
Hadron1776
ItsNickBarry
Janaszar
jmhorjus
jostephd
jozef-mitro
kozbot
lheckemann
MageKing17
MarcelineVQ
mattsoulanille
MCOfficer
Nescio0
NomadicVolcano
oo13
OverYrPaygrade
petervdmeer
pscamman
Rakete1111
reizbar
rlane
rzahniser
sdennie
sfiera
tboby
thomasballinger
TurkeyMcMac
willbeason
zwparchman
</details>

File: credits.txt [hunk 10]
```diff
@@ -61,3 +79,188 @@ Beta Testers
...
+  MarcelineVQ
+  MasterOfGrey
+  mattsoulanille
+  maxrd2
+  MCOfficer
+  MessyMix
+  MillerNerd
+  mimirzero
+  mootootwo
+  mybuddymichael
```
[the remark this answers is not in the document]
Comment (Zitchas) [reply credits.txt:203]: Quite possible, but unless we know for sure, I'm inclined to just leave it in.

File: credits.txt (hunk 9 above)
Comment (Zitchas) [reply credits.txt:20]: Agreed. That being said, exactly what counts as "programming" I think is probably up to you Dev's; and tehhowch in particular.

As far as that goes, I think inclusion in specific sections like "Programming" or "Writers" or "Editors" should be something that we actively manage. So someone doesn't get put in "Writers" as soon as they've created one job. The category would be for people who have made (or are) making a continued and/or long-lasting impact on the story. Likewise for Programmers. If I make one small change, maybe add a single attribute somewhere, I don't think I'd merit inclusion on the list of programmers. If I improve my coding skills and start making regular contributions that are deemed beneficial/useful, though... 

Basically, what I'm getting at is that each of these categories should be places where we are highlighting people that have made a significant contribution to that area of the game.

File: credits.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Amazinite) [reply credits.txt:203]: The single commit from this author isn't actually tied to a GitHub account and wasn't from a PR, with the commit's author being MZ. I've seen a similar thing occur when I pushed a commit to a branch through the GitHub desktop app without actually being logged onto my account.

[the remark this answers is not in the document]
Comment (Zitchas) [reply credits.txt:203]: Ah, OK.

File: credits.txt (hunk 9 above)
Comment (Amazinite) [reply credits.txt:20]: Well if we're going to curate these lists to just major contributions, then I'd say the current writers list could drop some names, as I only suggested all those who had simply created missions instead of all those who had contributed something I'd consider significant (as here I've simply listed all those who've made changes to the source code instead of only listing significant contributors): Czartrak and MasterOfGrey only added a few jobs; MessyMix, Anarchist2, and Fzzr have only added one mission each; and Jana's contribution to the Nanachi missions could more appropriately be considered to fall under "Proof Readers & Editors".

Comment (Zitchas) [reply credits.txt:20]: OK, I removed those and moved Jana over to Editors.

I don't really know a whole lot about how things work on the programming side; if you thing it'd make more sense to have separate sections for new mechanics and major bug fixes, I'm fine with that. Just let me know who should go on each list.

File: credits.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zitchas) [reply credits.txt:203]: OK, removed.

File: credits.txt (hunk 7 above)
Comment (Zitchas) [reply credits.txt:119]: Thanks!

File: credits.txt (hunk 9 above)
Comment (Amazinite) [reply credits.txt:20]: <details><summary>Here's a shortened list from the one above, with reasons for each.</summary>

AdamKauffman (#4472 was sizable.)
colbygallup (#772 You can thank him for piercing weapons.)
comnom (Various fixes/changes.)
CyberShadow (#3727 made zooming better and I like that.)
db47h (Various fixes/changes.)
Elyssaen (Various fixes/changes.)
EndrosG (" ")
flaviojs (" ")
guillochon (" ")
Hacklin (#2982 is responsible for how movement works right now. Might not be enough for others, but given #4924, I realize #2982 has actually been fairly impactful.)
Hadron1776 (Various fixes/changes,)
ItsNickBarry (#922 #923 responsible for solar collection and solar panel outfits.)
kozbot (Various fixes/changes.)
MarcelineVQ (" ")
NomadicVolcano (" ")
oo13 (#4538)
petervdmeer (#4370 ongoing auto testing work.)
tehhowch (Self-explanatory.)
thomasballinger (#3828 improved zoom further. Also, wholesome tweet.)

When it came to "various fixes/changes," I basically looked for if they'd contributed more than just one or two minor changes, so if I gave that as the reasoning then that person has contributed at least a handful of changes big and small. We can trim this list down further too if we want; I'd perhaps drop the names of those where my reasoning is that they contributed a single, small mechanic/change, e.g. colbygallup, Hacklin, ItsNickBarry, thomasballinger. I've already omitted pscamman, who I considered for having done #3802, but that's only a single two line change, so I don't think it met the bar.
</details>

Comment (Zitchas) [reply credits.txt:20]: OK. I have renamed the category "Major Programming" and replaced the previous list with this new one, with some names removed (the ones you listed there as being small changes)

Also renamed the "Missions & Story Creators" to "Story Creators" because I think that's a bit more clear that these aren't just the people who have made a single mission or two; they're actually writing stories.

File: credits.txt [hunk 11]
```diff
@@ -5,12 +5,35 @@ This game is open source.
...
 artwork, or programming, visit:
-https://endless-sky.github.io
-The game's Steam forums
-Discord (invite: ZeuASSx)
+- https://endless-sky.github.io
+- The game's Steam forums
+- Discord (invite: ZeuASSx)
 
-Programming
+Creator
```
Comment (tehhowch) [inline credits.txt:12]: Creator and Lead Programmer

File: credits.txt [hunk 12]
```diff
@@ -61,3 +84,187 @@ Beta Testers
   Sean Fahey
   Mauricio Gomes
   Iaz Poolar
+
+Special thanks to MCOfficer for
+creating, managing, and hosting
+the nightly builds that make
+so much of this work possible.
+
+Proof Readers & Editors
```
Comment (tehhowch) [inline credits.txt:93]: Editors & Proofreaders

File: credits.txt [hunk 13]
```diff
@@ -61,3 +84,187 @@ Beta Testers
...
+  Janaszar
+  LocalGod79
+  MCOfficer
+  Moonbeam
+  RecursiveParadox
+  Tadrix
+  temtemy
+  Zearth Goh aka TheMarksman
+
+Story Creators
```
Comment (tehhowch) [inline credits.txt:107]: Storyline & Faction Authors

File: credits.txt (hunk 11 above)
Comment (Zitchas) [reply credits.txt:12]: Sounds good. Done.

File: credits.txt (hunk 12 above)
Comment (Zitchas) [reply credits.txt:93]: That does sound better. Made the change.

File: credits.txt (hunk 13 above)
Comment (Zitchas) [reply credits.txt:107]: That's definitely a bit more clear. Changed!

File: credits.txt [hunk 14]
```diff
@@ -61,3 +84,187 @@ Beta Testers
   Sean Fahey
   Mauricio Gomes
   Iaz Poolar
+
+Special thanks to MCOfficer for
+creating, managing, and hosting
+the nightly builds that make
```
Comment (MCOfficer) [inline credits.txt:90]: might wanna change that to continuous, since that is supposed to replace the nightlies.

Comment (Zitchas) [reply credits.txt:90]: Hmmm... I'm inclined to leave it like this until the continuous stuff is fully implemented and we're confident in it. But yeah, definitely something to change once we've made the shift. (yes, I know it *looks* up and running, but it is early yet)

Comment (MCOfficer) [reply credits.txt:90]: As far as i'm concerned, the continuous builds are already more stable than the nightlies could ever be.
The OSX version has "proper" library paths, and the windows version is built on windows using the official makefile, not some half-assed substitute. Plus there are 2 linux versions which the nightlies didn't have to begin with.

Comment (Zitchas) [reply credits.txt]: And yet, your server's only ever been down once when I needed it - and GitHub seems to fail at stuff on a regular basis...

That being said, I changed it to say "nightly & continuous" in front of builds.
